File: 백준/15685_드래곤커브/Main.java
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

class Solution {

    static class Pos {
        int r, c;

        Pos(int r, int c) {
            this.r = r;
            this.c = c;
        }

        public Pos add(Pos pos) {
            return new Pos(r + pos.r, c + pos.c);
        }
    }

    private static final Pos[] dirs = {
        new Pos(0, 1),
        new Pos(-1, 0),
        new Pos(0, -1),
        new Pos(1, 0)
    };
    private int N;
    private boolean[][] board;
    private int[][] curves;

    public Solution(BufferedReader reader) throws IOException {
        N = Integer.parseInt(reader.readLine());
        board = new boolean[102][102];
        curves = new int[N][4];

        for(int i = 0 ; i < N ; i++) {
            String[] tokens = reader.readLine().split(" ");
            int c = Integer.parseInt(tokens[0]);
            int r = Integer.parseInt(tokens[1]);
            int d = Integer.parseInt(tokens[2]);
            int g = Integer.parseInt(tokens[3]);

            curves[i][0] = c;
            curves[i][1] = r;
            curves[i][2] = d;
            curves[i][3] = g;
        }
    }

    private Pos rotate(Pos target, Pos pivot) {
        int dr = target.r - pivot.r;
        int dc = target.c - pivot.c;
        int nc = pivot.c - dr;
        int nr = pivot.r + dc;

        return new Pos(nr, nc);
    }

    private void makeDragonCurve(int idx) {
        List<Pos> points = new ArrayList<>();

        Pos start = new Pos(curves[idx][1], curves[idx][0]);
        Pos end = start.add(dirs[curves[idx][2]]);
        int gen = curves[idx][3];

        points.add(start);
        points.add(end);

        for(int i = 0 ; i < gen ;i++) {
            Pos pivot = points.get(points.size() - 1);
            List<Pos> newPoints = new ArrayList<>();

            for(int j = points.size() - 2; j >= 0 ; j--) {
                Pos newPoint = rotate(points.get(j), pivot);
                newPoints.add(newPoint);
            }

            points.addAll(newPoints);
        }

        for(Pos point : points) {
            board[point.r][point.c] = true;
        }
    }

    private int countSquares() {
        int count = 0;
        for(int i = 0 ; i < 100 ; i++) {
            for(int j = 0;  j < 100 ; j++) {
                if(board[i][j] && board[i][j + 1] && board[i + 1][j] && board[i + 1][j + 1]) {
                    count++;
                }
            }
        }

        return count;
    }

    public void run() {
        StringBuilder sb = new StringBuilder();

        for(int i = 0 ; i < N ; i++) {
            makeDragonCurve(i);
        }
        sb.append(countSquares());

        System.out.println(sb.toString());
    }
}

public class Main {

    static void testRun(int no) throws IOException {
        Path base = Paths.get("");
        String testFileDirName = base.toAbsolutePath() + "/src/main/resources/examples/" + String.valueOf(no);

        File dir = null;
        try {
            dir = new File(testFileDirName);
        }catch(NullPointerException e) {
            return;
        }
        File[] files = dir.listFiles();

        if(files == null) {
            return;
        }

        for(int i = 0 ; i < files.length ; i++){
            String fileName = files[i].getName();
            String fullPath = testFileDirName + "/" + fileName;
            System.out.println("Test file name : " + fullPath);
            BufferedReader reader = new BufferedReader(new FileReader(fullPath));
            reader.mark(262144);
            List<String> lines = reader.lines()
                    .collect(Collectors.toList());

            boolean found = true;
            for(String line : lines) {
                if(line.contains("answer")) {
                    found = true;
                }

                if(found) {
                    System.out.println(line);
                }
            }
            reader.reset();

            System.out.println("answer ");
            new Solution(reader).run();
            reader.close();
        }
    }

    static void run() throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        new Solution(reader).run();
        reader.close();
    }

    public static void main(String[] args) throws IOException {
        testRun(15685);
        //run();
    }
}

File: 백준/15685_드래곤커브/main.rs
trait Solution {
    fn new()->Self ;
    fn run(&mut self);
}

struct CurveInfo {
    x: i32,
    y: i32,
    d: i32,
    g: i32
}

struct Pos {
    x: i32,
    y: i32
}

struct Problem {
    n: i32,
    curves: Vec<CurveInfo>
}

const dx: [i32; 4] = [1, 0, -1, 0];
const dy: [i32; 4] = [0, -1, 0, 1];

impl Solution for Problem {

    fn new() -> Self {
        let n: i32 = IO::read_line().parse().unwrap();
        let mut curves: Vec<CurveInfo> = Vec::new();
        for i in 0..n {
            let tokens: Vec<i32> = IO::read_line()
                .split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect();
            curves.push( CurveInfo {
                x: tokens[0],
                y: tokens[1],
                d: tokens[2],
                g: tokens[3]
            });
        }

        Self {
            n : n,
            curves: curves
        }
    }

    fn run(&mut self) {
        let mut board: Vec<Vec<bool>> = vec![vec![false; 101]; 101];

        for curve in &self.curves {
            make_curves(&curve, &mut board);
        }

        let answer = count_squares(&board);
        println!("{}",answer);
    }
}

fn count_squares(board: &Vec<Vec<bool>>) -> i32 {
    let mut count: i32 = 0;

    for r in 0..100 {
        for c in 0..100 {
            if board[r][c] && board[r+1][c] && board[r][c+1] && board[r+1][c+1] {
                count+=1;
            }
        }
    }

    count
}

fn rotate(x: i32, y: i32, px: i32, py: i32) -> (i32, i32) {
    let tx: i32 = x - px;
    let ty: i32 = y - py;
    // 0 -1  tx
    // 1  0  ty

    ( -ty + px, tx + py)
}

fn make_curves(curve: &CurveInfo, board: &mut Vec<Vec<bool>>) {
    let mut trace: Vec<Pos> = Vec::new();

    trace.push(Pos{
        x: curve.x,
        y: curve.y
    });

    trace.push(Pos{
        x: curve.x + dx[curve.d as usize],
        y: curve.y + dy[curve.d as usize]
    });


    for gen in 0..curve.g {
        let pivot = &trace[trace.len() - 1];
        let mut new_points: Vec<Pos> = Vec::new();
        for point in &trace {
            if point.x == pivot.x && point.y == pivot.y {
                continue;
            }
            let np = rotate(point.x, point.y, pivot.x, pivot.y);
            new_points.push(Pos {
                x: np.0,
                y: np.1
            });
        }
        new_points.reverse();
        trace.extend(new_points);
    }

    //println!("curves x, y = {}, {}, gen {} , dir {}", curve.x, curve.y, curve.g, curve.d);

    for point in &trace {
        //println!("  point {} {}", point.x, point.y);
        if(point.y > 100 || point.x > 100) { continue; }
        board[(point.y) as usize][(point.x) as usize] = true;
    }
}

fn main() {
    let mut problem: Problem = Problem::new();
    problem.run();
}

mod IO {
    pub fn read_line() -> String {
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        buf.trim().to_string()
    }

    pub fn read_copy_line(buf: &mut String) {
        std::io::stdin().read_line(buf).unwrap();
    }
}
